blog/2024-05-21-blog-status-of-lbse-in-webkit.md

```diff
@@ -0,0 +1,157 @@
+
+## Cycle detection
+
+It is quite easy in SVG to cause direct circular references:
+
+
+```html
+<svg>
+    <defs>
+        <pattern id="p" xlink:href="#p" />
+    </defs>
+    ...
+</svg>
+```
+
+It is also possible to cause indirect circular references:
+
+```html
+<svg>
+    <defs>
+        <mask id="z" />
+            <rect mask="url(#z)" />
+        </mask>
+    </defs>
+    <ellipse mask="url(#z)" />
+</svg>
+```
+
+We have in the past solved this in the legacy engine in an ad-hoc way in various places in the engine; it tried to break cycles before rendering, but still needed cycle protections in various places, since the solution was never unified or complete.
+
+In **February 2024** we provided a unified solution for LBSE by introducing `SVGVisitedRendererTracking`, see this <a href="https://commits.webkit.org/274392@main">commit</a>. In the new approach, we don't attempt to remove cycles, but detect them everywhere upon usage and stop processing in well-defined ways, all centralized in <code>SVGVisitedRendererTracking</code>.
+
+## Nested mask/pattern slowness
+
+In **April 2024** we addressed slowness problems with nested masks/patterns. As an example, consider this for nested masks:
+
+```html
+<svg>
+    <defs>
+        <mask id="z" />
+            <rect id="1" mask="url(#y)" />
+            <rect id="2" mask="url(#y)" />
+            ...
+        </mask>
+        <mask id="y" />
+            <rect id="1" mask="url(#x)" />
+            <rect id="2" mask="url(#x)" />
+            ...
+        </mask>
+        ...
+    </defs>
+    <ellipse mask="url(#z)" />
+</svg>
+```
+
+For above example, the complexity can be increased at will by adding more masks and contents per mask.
+
+The solution was twofold:
+
+- for masks, we realized bounding box calculations for a mask were not affected by masks used in the mask contents, so we could cut off bounding box calculations for nested masks.
+- for both masks and patterns, we added caching of image buffers per resource client so nested masks/patterns that are already encountered can reuse the image buffer cache.
+
+See optimizations here for nested <a href="https://commits.webkit.org/273820@main">masks</a> and <a href="https://commits.webkit.org/277306@main">patterns</a>.
+
+## Next steps
+
+For the short and mid-term, the plan is to make LBSE at least as good as legacy in regards to test coverage, i.e. no test should pass in legacy but not in LBSE. We have made a lot of progress over the
+last 7 months just because of the amount of SVG resources that were implemented, but for example we will need to have SVG filters in place to pass this goal.
+
+Another goal is to make sure LBSE qualifies for all security requirements, we are already adopting a lot of good <a href="https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines">smart pointer practices</a> to support this.
```

Comment:
This reads a bit strange to me too... sounds like there's something missing between the comma and the "we are".

How about this:

> Another goal is to make sure LBSE qualifies for all security requirements, as that would be a blocker otherwise before the current engine can be replaced. Fortunataly, we are already taking this into account in several ways to support this, such as  adopting a lot of good <a href="https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines">smart pointer practices</a>, for instance.